### PR TITLE
Gracefully handle "Cannot read property .match of undefined"

### PR DESCRIPTION
**Pull request by danielleadams**
# What / Why
There's a bug in which an npm command will raise an error `Cannot read property .match on undefined` and exit the script with an error. There are a few open issues:
- https://github.com/npm/cli/issues/973 (occurs with `npm install`)
- https://github.com/npm/cli/issues/961 (occurs with `npm install`)
- https://github.com/npm/cli/issues/703 (occurs with `npm audit fix`)
- https://github.com/npm/cli/issues/403 (occurs with `npm uninstall --save-dev`)

I've run into this with `npm prune`, but I was unable to reproduce it while looking into it. This doesn't fix the root of the issue, which it appears that empty parameters are being generated in the lock file something like this:

```
"package-a": {
  "dependencies": {
    "package-b": {}
  }
}
```

And causing an undefined error here: https://github.com/npm/cli/blob/latest/lib/install/inflate-shrinkwrap.js#L87

Because there is no `"version"` key that is passed in here: https://github.com/npm/cli/blob/latest/lib/install/inflate-shrinkwrap.js#L99 (which is returning an `undefined`).

## Solution
This catches a dependency if it's empty and displays an error message to the user to regenerate the package-lock.json or npm-shrinkwrap.json.

## TODO
- ~write tests~
- split out the messaging based on the lockfile type if requested

## References

* n/a

### CODE REVIEW

**Review by ruyadorno — APPROVED** [review]
awesome! 🎉 thanks @danielleadams

**Comment by ruyadorno** [comment]
Sorry for the delay @danielleadams I was meant to write back to you after the last release - I tried landing this in `v6.14.5` but found out that there's some broken tests, basically it seems like the commit got the green light because CI just skipped all tests last time 😭

File: lib/install/inflate-shrinkwrap.js
Comment (ruyadorno) [inline lib/install/inflate-shrinkwrap.js]: I believe you're going to have to return a `Promise` here in order to not break the promise chain, maybe simply?

```diff
-      return errorHandler(message)
+      return Promise.reject(errorHandler(message))
```

**Review by ruyadorno — CHANGES_REQUESTED** [review]
Tests are currently failing

**Comment by ruyadorno** [comment]
btw, if you rebase on top of current `latest` branch it should fix the problem with CI 😊

**Comment by danielleadams** [comment]
oh whoops! Ok.. let me rebase and see if it fixes, or otherwise fix it.

**Review by ruyadorno — APPROVED** [review]
awesome 🎉 thanks again @danielleadams 😊

**Review by ruyadorno — CHANGES_REQUESTED** [review]
digging a bit more into the output from each of the CI targets, there's actually a bunch of silent failures in the form of:

```
 npm ERR! cb() never called!

npm ERR! This is an error with npm itself. Please report this error at:
npm ERR!     <https://npm.community>

npm ERR! A complete log of this run can be found in:
npm ERR!     C:\npm\cache\_logs\2020-06-11T20_02_29_720Z-debug.log
```

which seems to imply the broken chain of promise is still an issue 😞

**Comment by danielleadams** [comment]
@ruyadorno I made changes to how errors are handling... instead of importing and using the error handler, I returned a failed promise with an error from the if statement. The tests pass, which tells me that the error is passed up the call stack and still handled. Let me know what you think 👍

**Review by ruyadorno — APPROVED** [review]

**Pull request closed without merging**
